scripts/ioc_deploy.py

```diff
@@ -301,6 +426,143 @@ def make_in(deploy_dir: str, dry_run: bool) -> int:
         return subprocess.run(["make"], cwd=deploy_dir).returncode
 
 
+def set_permissions(deploy_dir: str, allow_write: bool, dry_run: bool) -> int:
+    """
+    Apply or remove write protection from a deploy repo.
+
+    You may or may not have permissions to do this to releases created by other users.
+
+    allow_write=True involves adding "w" permissions
+    for the owner and for the group. "w" permissions will never be added for other users.
+    We will also add write permissions to the top-level directory.
+
+    allow_write=False involves removing the "w" permissions
+    from all files, and from directories that contain untracked files.
+    We will also remove permissions from the top-level direcotry.
+    """
+    if dry_run and not os.path.isdir(deploy_dir):
+        # Dry run has nothing to do if we didn't build the dir
+        # Most things past this point will error out
+        logger.info("Dry-run: skipping permission changes on never-made directory")
+        return ReturnCode.SUCCESS
+
+    # Ignore these directories
+    exclude = (".git",)
+
+    if allow_write:
+        # Lazy and simple: chmod everything
+        perms = get_allow_write_rule(deploy_dir)
+        if dry_run:
+            logger.info(f"Dry-run: skipping chmod({deploy_dir}, {oct(perms)})")
+        else:
+            os.chmod(deploy_dir, perms)
+        for dirpath, dirnames, filenames in exclude_walk(deploy_dir, exclude=exclude):
+            for name in dirnames + filenames:
+                full_path = os.path.join(dirpath, name)
+                perms = get_allow_write_rule(full_path)
+                if dry_run:
+                    logger.debug(f"Dry-run: skipping chmod({full_path}, {oct(perms)})")
+                else:
+                    logger.debug(f"chmod({full_path}, {oct(perms)})")
+                    os.chmod(full_path, perms)
+        return ReturnCode.SUCCESS
+
+    # Compare the files that exist to the files that are tracked by git
+    try:
+        ls_files_output = subprocess.check_output(
+            ["git", "-C", deploy_dir, "ls-files"],
+            universal_newlines=True,
+        )
+    except subprocess.CalledProcessError as exc:
+        return exc.returncode
+    deploy_path = Path(deploy_dir)
+    tracked_paths = [
+        deploy_path / subpath.strip() for subpath in ls_files_output.splitlines()
+    ]
+    build_dir_paths = set()
+
+    def accumulate_subpaths(subpath: Path):
+        for path in subpath.iterdir():
+            if path.is_symlink():
+                continue
+            elif path.name in exclude:
+                continue
+            elif path.is_dir():
+                accumulate_subpaths(path)
+            elif path.is_file():
+                if path not in tracked_paths:
+                    build_dir_paths.add(str(path.parent))
```

Comment:
I think I understand this now.  If we come across an untracked file, we add its direct parent to `build_dir_paths`, and that folder gets write protected.  This works for structures like our existing bin folders:
```
bin/   (ignored parent of protected folder)
└── rhel7-x86_64  (protected folder)
    └── ims  (untracked file)

1 directory, 1 file
```

If someone breaks this pattern by accidentally dropping a file in `bin/`, this script breaks right?  (an action which is not explicitly forbidden)

Reply:
You're correct, many sorts of manual changes to the build directory will confuse the script and give unexpected outputs. That might be enough of a reason to pick a different methodology.

Reply:
fwiw I'm leaning towards simplifying the whole process, write protecting all files and directories with no exceptions, and requiring editing the permissions (using the script) if someone wants to make a new build.

Reply:
I wonder if the simpler (but more brittle) solution is to just add folders like `bin/` to a "folder-exclude" list, but then write protect the contents of these directories.  

The `make` processes may change, but we can probably make some guarantees about what folders the Makefile touches.

Reply:
> write protecting all files and directories with no exceptions, and requiring editing the permissions (using the script) if someone wants to make a new build. 
I can get on board with this.  Even though multi-arch builds are the exception, using the script still provides a path forward

Reply:
@slactjohnson do you have any opinions about this specific topic? Making the permissions handling simpler vs leaving them as-is vs changing how it works for these sorts of edge cases?

Reply:
@ZLLentz I think I'm in favor of keeping things simpler, at least to start. I think that it's possible that we have more multi-arch shenanigans going on in the future than we have in the past as we manage the three body problem of rhel5 <-> rhel7 <-> rocky9, but I think if that becomes painful we can re-visit.

Reply:
OK, we are largely in agreement then. I will make the handling maximally simple alongside the other adjustments. The old code for more specific targeting will still exist in git history land if we need to bring it back.